lib/internal/webstreams/adapters.js

```diff
@@ -0,0 +1,746 @@
+
+const {
+  ArrayPrototypeMap,
+  PromiseAll,
+  PromisePrototypeThen,
+  PromisePrototypeFinally,
+  Uint8Array,
+} = primordials;
+
+const {
+  ReadableStream,
+  isReadableStream,
+} = require('internal/webstreams/readablestream');
+
+const {
+  WritableStream,
+  isWritableStream,
+} = require('internal/webstreams/writablestream');
+
+const {
+  ByteLengthQueuingStrategy,
+  CountQueuingStrategy,
+} = require('internal/webstreams/queuingstrategies');
+
+const {
+  setPromiseHandled,
+} = require('internal/webstreams/util');
+
+const {
+  Writable,
+  Readable,
+  Duplex,
+} = require('stream');
+
+const {
+  Buffer,
+} = require('buffer');
+
+const {
+  errnoException,
+  codes: {
+    ERR_INVALID_ARG_TYPE,
+    ERR_INVALID_ARG_VALUE,
+    ERR_INVALID_STATE,
+  },
+} = require('internal/errors');
+
+const {
+  createDeferredPromise,
+} = require('internal/util');
+
+const {
+  validateAbortSignal,
+  validateBoolean,
+  validateObject,
+} = require('internal/validators');
+
+const {
+  WriteWrap,
+  ShutdownWrap,
+  kReadBytesOrError,
+  kLastWriteWasAsync,
+  streamBaseState,
+} = internalBinding('stream_wrap');
+
+const { UV_EOF } = internalBinding('uv');
+
+function newWritableStreamFromStreamWritable(streamWritable) {
+  if (typeof streamWritable?._writableState !== 'object') {
+    throw new ERR_INVALID_ARG_TYPE(
+      'streamWritable',
+      'stream.Writable',
+      streamWritable);
+  }
+
+  const highWaterMark = streamWritable.writableHighWaterMark;
+  const strategy =
+    streamWritable.writableObjectMode ?
+      new CountQueuingStrategy({ highWaterMark }) :
+      new ByteLengthQueuingStrategy({ highWaterMark });
+
+  let controller;
+  let backpressurePromise;
+  let stream;
+  let closed = false;
+  let errored = false;
+
+  function onClose() {
+    closed = true;
+    streamWritable.off('close', onClose);
+    streamWritable.off('drain', onDrain);
+    streamWritable.off('error', onError);
+    setPromiseHandled(stream.close());
+  }
+
+  function onDrain() {
+    if (backpressurePromise !== undefined)
+      backpressurePromise.resolve();
+  }
+
+  function onError(error) {
+    errored = true;
+    if (backpressurePromise !== undefined)
+      backpressurePromise.reject(error);
+    controller.error(error);
+  }
+
+  streamWritable.once('close', onClose);
+  streamWritable.once('error', onError);
+  streamWritable.on('drain', onDrain);
+
+  stream = new WritableStream({
+    start(c) { controller = c; },
+
+    async write(chunk) {
+      if (!streamWritable.write(chunk)) {
+        backpressurePromise = createDeferredPromise();
+        return PromisePrototypeFinally(
+          backpressurePromise.promise, () => {
+            backpressurePromise = undefined;
+          });
+      }
+    },
+
+    abort(reason) {
+      if (!streamWritable.destroyed && !errored)
+        streamWritable.destroy(reason);
+    },
+
+    close() {
+      if (!streamWritable.writableEnded && !closed)
+        streamWritable.end();
```

Comment:
This should wait for the `finish` event, to be sure that *all* data has been flushed. A chunk that hasn't been flushed yet might still error the `streamWritable`, so the `WritableStream` must not be closed yet.